### PR TITLE
QNTM-711: Serialize/Deserialize Dynamo Preferences from JSON

### PR DESCRIPTION
**Pull request by ColinDayOrg**
### Purpose

This PR serializes/deserializes Dynamo-specific preferences from JSON

The Dynamo preferences addressed in this PR are;
- ScaleFactor
- HasRunWithoutCrash
- IsVisibleInDynamoLibrary
- Version
- RunType
- RunPeriod

### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [n/a] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [n/a] User facing strings, if any, are extracted into `*.resx` files
- [X] All tests pass using the self-service CI.
- [n/a] Snapshot of UI changes, if any.
- [n/a] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner

### CODE REVIEW

File: src/DynamoCore/Models/DynamoModel.cs [hunk 1]
```diff
@@ -1299,6 +1328,28 @@ public void OpenFileFromPath(string filePath, bool forceManualExecutionMode = fa
             Logger.LogError("Could not open workspace at: " + filePath);
         }
 
+        static private DynamoPreferencesData DynamoPreferencesDataFromJson(string json)
+        {
+            JsonReader reader = new JsonTextReader(new StringReader(json));
+            var obj = JObject.Load(reader);
+            var viewBlock = obj["View"];
+            var dynamoBlock = viewBlock["Dynamo"];
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: let's assume that for some graphs this value is null - we should still deserialize the rest of the graph okay, currently it looks like we'll throw immediately on calling ToString() below, and then in a bunch of places during OpenJSON file - can we check for null in all these places?

File: src/DynamoCore/Models/DynamoModel.cs [hunk 2]
```diff
@@ -1449,13 +1494,33 @@ private bool OpenJsonFile(WorkspaceInfo workspaceInfo, string fileContents, out
                 false,
                 CustomNodeManager);
 
-            workspace.FileName = workspaceInfo.FileName;
+            workspace.FileName = filePath;
+            workspace.ScaleFactor = dynamoPreferences.ScaleFactor;
```
Comment (mjkkirschner) [inline src/DynamoCore/Models/DynamoModel.cs]: lets harden this as described above.  - a potential solution to keep this code free of null checks, is that if the deserialization of dynamo block fails, fill it up with default values?

File: src/DynamoCoreWpf/ViewModels/Core/WorkspaceViewModel.cs [hunk 3]
```diff
@@ -182,6 +181,41 @@ private void OnRequestShowInCanvasSearch(object param)
...
+              bool isVisibleInDynamoLibrary = true;
+              CustomNodeWorkspaceModel customNodeWorkspace = Model as CustomNodeWorkspaceModel;
+              if (customNodeWorkspace != null)
+                isVisibleInDynamoLibrary = customNodeWorkspace.IsVisibleInDynamoLibrary;
+
+              return new DynamoPreferencesData(
+                Model.ScaleFactor,
+                hasRunWithoutCrash,
+                isVisibleInDynamoLibrary,
+                Model.WorkspaceVersion.ToString(),
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/WorkspaceViewModel.cs]: hmmm, I think when we save we actually want to save the current DynamoModel version not the old workspace version right?

File: src/DynamoCore/Models/DynamoModel.cs (hunk 1 above)
Comment (gregmarr) [reply src/DynamoCore/Models/DynamoModel.cs]: Neither `View` nor `Dynamo` are required, and so we have to assume that they may be null.

**Comment by mjkkirschner** [comment]
@ColinDayOrg I think we need to handle the case where the entire dynamoPreferences object is missing from the view block, perhaps even the entire view.

also, can we add tests for this to the serialization tests?

File: src/DynamoCore/Models/DynamoModel.cs (hunk 1 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Models/DynamoModel.cs]: I have added code to handle both cases where either the view block is entirely missing, or there is a view block but no Dynamo preferences block in the view block.

File: src/DynamoCore/Models/DynamoModel.cs (hunk 2 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Models/DynamoModel.cs]: I have added code to set default preferences in the case that the Dynamo preferences block is not present.

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs [hunk 4]
```diff
@@ -1566,6 +1566,9 @@ protected void Log(string msg, WarningLevel severity)
         /// <param name="workspaceViewInfo">The extra view information from the workspace to update the model with.</param>
         public void UpdateWithExtraWorkspaceViewInfo(ExtraWorkspaceViewInfo workspaceViewInfo)
         {
+            if (workspaceViewInfo == null)
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: so if we just return will the view default to something sensible?

**Comment by mjkkirschner** [comment]
@ColinDayOrg fixes look good, we can merge this now - now that I think about it, seems to add tests we'll need to wait until we have WPF tests for the view.

**Comment by mjkkirschner** [comment]
@ColinDayOrg one question and LGTM

File: src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs (hunk 4 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs]: With no view block the code will skip the section where the view block values are used to modify the loaded workspace. This results in the position and zoom of the view defaulting to 0,0 and 1, all the nodes being crunched together, and no notes or annotations, but this is to be expected with no view information. I can't really say that this is "sensible" per se, but I don't think that there is anything to really be done about it (right now at least).

I will need to do some more testing to determine the specifics, but I have tested with both no view block at all and with a view block and no Dynamo preferences and there are no crashes. In the case of no Dynamo preferences block there are some default values that will be used instead.

**Comment by mjkkirschner** [comment]
@ColinDayOrg cool, I just want to make sure we didn't crash.

**Comment by ColinDayOrg** [comment]
I am just about to add a change to save the file using the current version instead of the version that was loaded (good catch btw). I was then going to look into the testing aspect to see if there is some testing we can add at the workspace level, but I agree, the testing will also need to be at the view level as well. Do you think that I should continue looking at the testing we can do now, or wait for the view level tests? I could also just merge the current implementation for now before looking into the tests, and then continue on looking into the tests... thoughts?

**Comment by mjkkirschner** [comment]
fix the version stuff and lets merge this in.

**Comment by ColinDayOrg** [comment]
Ok, sounds good to me. I'm just running tests now one more time before merging.

**Pull request merged**